VCards in entry forms, list views, search indexing and visualization

litvinovg opened the pull request:
**[VIVO GitHub issue](https://github.com/vivo-project/VIVO/issues)**

Vitro PR

# What does this pull request do?
Added option to store participants as VCards for authorships, editorships and project participants.
Added option in editorship entry form to select organization as editor.
Added search indexing extension to index VCard participants.
Support visualization of VCard co-authors 

# What's new?
Modified entry forms:

- [AddAuthorsToInformationResourceGenerator.java](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-24828e4250b55f0bd838f2b4cf335ad0be4149856d104ce324b5535dca5a6f59)
[addAuthorsToInformationResource.ftl](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-7ac7aaa1813e85b9bce8800ee7af29892e0fcfd452e46a81f770e1e2bdde2de4)
[addAuthorsToInformationResource.js](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-98fa61faa54dc99ec927b0e1cb1f7b01ece763925bac2045fab7a74fbdaaf0e3)


- [AddEditorsToInformationResourceGenerator.java](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-4b674c4ebbcfb97820680db4977bd783f2eed4fe8ae71e865ca69e1ab9dd281c)
[addEditorsToInformationResource.ftl](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-6903fa17d6e69f5f38be0ed24f4ae9c152c03156f87ab6fd626b95976568a6b6)
[addEditorsToInformationResource.js](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-6932b147906da40aea93b38137090cffe159ac3dcee9fba836216e75efbaa988)

- [ProjectHasParticipantGenerator.java](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-47b050c9e3ce5791e9ff5503261effafef54d181654d19b5b9fc9400c0bfb8cf)
[projectHasParticipant.ftl](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-c4fe565f87beefb0a260c6028f8d5c3c0bb7d69ebbfa890be158898d63de70e4)

Created list view for presentation participant [listViewConfig-relatedExternalSpeakerRole.xml](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-bd42ade7cac72582396fc0b6478df91991782f326b9c4315e000ad65e052aa98)
Applied list view for participant faux property [PropertyConfig.n3](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-900c38f157ef62660ae2309859e95a712a0db6d5e1608cc707a33811bc4ef5fc)

Modified editorship list view to support VCards [listViewConfig-informationResourceInEditorship.xml](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-234844d4bc50259e97b512fa0ee38e05edf1430f864ca287e8620bb85fcce8cf) and [propStatement-informationResourceInEditorship.ftl](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-3611e6e63caa5ff247a91cfe1679658af6a8a2877a08baa76f31a8a2ee0bea2a)

VCard visualization modifications:
[Collaborator.java](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-9a5d7b4c2314be2b76e41c14703527f4bf7ee66db95fc2ae333d4aee5bd772a6)
[CollaborationDataViewHelper.java](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-081adcc6832f1148d23a05b780cadf5961453dedf6e649bbecb30ba6db182097)
[coAuthorPersonLevelD3.ftl](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-268a22d63ec57602f7b6098759008f66d7bf567f7272057dccb92c7edaaa764a)
[collaboratorToActivityCountTable.ftl](https://github.com/vivo-project/VIVO/compare/main...litvinovg:VIVO:vcards-original#diff-b69089529356d3691727d7e996b285658af45d96b14b463aaf20c3bbece13100)

# How should this be tested?
A description of what steps someone could take to:
* Try adding/removing participants to a publication by using authors, editors faux property:
* Add VCard participant
* Add person
* Select person from existing
* Add organization
* Select organization from existing
* For each case verify that added participant is listed in related list view
* Try adding participants to a project and presentations: vcard and real
* In person co-author network verify that vcard co-authors are visible if "Include external co-authors" is checked 
* Search for VCard participants

# Interested parties
@VIVO-project/vivo-committers

# Reviewers' expertise

Candidates for reviewing this PR should have some of the following expertises:
1. Java
2. HTML, CSS, JavaScript
3. FreeMarker
4. SPARQL
1. Natural language knowledge
    1. English
    2. German
    3. Spanish
    4. French
    5. Portuguese
    6. Russian
    7. Serbian

# Reviewers' report template
## General comment
A reviewer should provide here comments and suggestions for requested changes if any.
## Testing
A reviewer should briefly describe here how it was tested
## Code reviewing
A reviewer should briefly describe here which part was code reviewed

tawahle reviewed (changes requested):
In my opinion, an explanation for Vcard is required. Maybe something like this: "Please select VCARD only, when the person has not and is not supposed to have a profile in VIVO. Otherwise choose the type "Person"."

litvinovg commented:
> In my opinion, an explanation for Vcard is required. Maybe something like this: "Please select VCARD only, when the person has not and is not supposed to have a profile in VIVO. Otherwise choose the type "Person"."

@tawahle 
I modified description text as you suggested. Also fixed presentation participant form and list view.
Changes were redeployed on https://vivo.tib.eu/vivorc
Please test it once again when you have time.

tawahle reviewed (approved):
Everything is working fine.

chenejac commented on webapp/src/main/webapp/config/listViewConfig-informationResourceInEditorship.xml:
```diff
@@ -1,19 +1,20 @@
 <!-- $This file is distributed under the terms of the license in LICENSE$ -->
 
-<!-- See guidelines at https://wiki.lyrasis.org/display/VIVODOC112x/Custom+List+View+Configurations -->
+<!-- See guidelines at https://wiki.duraspace.org/x/eYXVAw -->
```
The link is broken? Why did you change the previous link?

chenejac replied:
this is also fixed. thanks

chenejac commented on home/src/main/resources/rdf/i18n/en_CA/interface-i18n/firsttime/vivo_UiLabel.ttl:
```diff
@@ -6266,3 +6266,17 @@ uil-data:degree_candidacy.VIVO
+uil-data:select_vcard_or_person.VIVO
+        rdf:type         owl:NamedIndividual ;
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Please select VCARD only, when the person doesn't have and is not supposed to have a profile in VIVO. Otherwise choose the type \"Person\"."@en-CA ;
```
Maybe we can use here: 

Please select external co-author, when the person doesn't have and is not supposed to have a profile in VIVO. Otherwise choose the type \"Person\"."

litvinovg replied:
Maybe. In that case VCARD option should be changed accordingly on entry form pages. In any case I think we should change all the translations altogether when there is an agreement on how it should look like to avoid changing the same thing multiple times. It was already changed once in accordance with this comment https://github.com/vivo-project/VIVO/pull/4015#pullrequestreview-2359604501

chenejac replied:
the new text looks good to me

chenejac commented on webapp/src/main/webapp/config/listViewConfig-relatedExternalSpeakerRole.xml:
```diff
@@ -0,0 +1,131 @@
+		
+
+        SELECT DISTINCT ?subclass
+						?vSubclass
```
please fix indentation

litvinovg replied:
Done

chenejac commented on webapp/src/main/webapp/config/listViewConfig-relatedExternalSpeakerRole.xml:
```diff
@@ -0,0 +1,131 @@
+    <query-select>
+        PREFIX rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt;
+        PREFIX core: &lt;http://vivoweb.org/ontology/core#&gt;
+		PREFIX foaf:  &lt;http://xmlns.com/foaf/0.1/&gt;
```
please fix indentation

litvinovg replied:
Done

chenejac commented on webapp/src/main/webapp/config/listViewConfig-relatedExternalSpeakerRole.xml:
```diff
@@ -0,0 +1,131 @@
+                    <precise-subquery>?subject ?property ?role .
+                    ?role a ?objectType .
+                    ?role &lt;http://purl.obolibrary.org/obo/RO_0000052&gt; ?indivInRole .
+					?indivInRole a vcard:Kind .</precise-subquery>
```
please fix indentation

litvinovg replied:
Done

chenejac commented on webapp/src/main/webapp/config/listViewConfig-relatedExternalSpeakerRole.xml:
```diff
@@ -0,0 +1,131 @@
+                ?role a ?objectType .</precise-subquery>
+                ?role &lt;http://purl.obolibrary.org/obo/RO_0000052&gt; ?indivInRole .
+                ?indivInRole a foaf:Person .
+				?indivInRole rdfs:label ?indivLabel .
```
please fix indentation

litvinovg replied:
Done

chenejac commented on webapp/src/main/webapp/templates/freemarker/edit/forms/js/addEditorsToInformationResource.js:
```diff
@@ -64,9 +64,20 @@ var addEditorForm = {
         this.personLink = $('a#personLink');
         this.returnLink = $('a#returnLink');
+		
+		this.orgSection.hide();
```
please fix indentation

litvinovg replied:
Done

chenejac commented on webapp/src/main/webapp/templates/freemarker/visualization/personlevel/coAuthorPersonLevelD3.ftl:
```diff
@@ -96,57 +96,82 @@ $(document).ready(function(){
     var labels = [];
     var uris   = [];
     var matrix = [];
+    var width  = 725;
+    var height = 725;
+    var padding = 175;
```
Could we have the issue with responsiveness because fixing those values?

litvinovg replied:
Not sure. But it is not something new here, take a look at old lines 122-124

chenejac replied:
ok, let's keep it as is in this PR, later we work on more responsive theme, we might improve this

chenejac reviewed (changes requested):
@litvinovg thanks for the PR. Looks nice and works well. I have posted a couple of comments. 

I am wondering whether we should add a middle class between VIVOBaseGenerator on side, and AddEditorsToInformationResourceGenerator and AddAuthorsToInformationResourceGenerator on the other side, to encapsulate shared code in that middle class (for instance it might be entitled AddContributorToInformationResourceGenerator).

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddAuthorsToInformationResourceGenerator.java:
```diff
@@ -109,6 +109,14 @@ public String getN3PrefixString() {
+        return "@prefix fail_pattern: ?createVCard .\n";
+    }
+
+    public String getDisableRealPersonPrefix() {
```
Could this method be named better? What does RealPerson indicate?

litvinovg replied:
Renamed method

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddAuthorsToInformationResourceGenerator.java:
```diff
@@ -171,13 +185,51 @@ private String getN3NewPersonLastName() {
 	}
 
+	// Changes here for creating vcards for external authors
+	private String getN3NewPersonVCardFirstName() {
```
The method naming here is also a bit confusing here with `getN3...`

litvinovg replied:
Yes. But it is aligned with other method names used in N3 generators.

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/AddAuthorsToInformationResourceGenerator.java:
```diff
@@ -436,7 +496,7 @@ public void addFormSpecificData(EditConfigurationVTwo editConfiguration, VitroRe
 	    + "	 bind ( COALESCE(?middleName, \"\") As ?middleName1) . \n"
 	    + "	 bind ( COALESCE(?lastName, \"\") As ?lastName1) . \n"
-	    + "	 bind (concat(str(?lastName1 + \", \"),str(?middleName1 + \" \"),str(?firstName1)) as ?authorName) . \n"
+	    + "	 bind (concat(str(str(?lastName1) + \", \"),str(str(?middleName1) + \" \"),str(?firstName1)) as ?authorName) . \n"
```
Are the nested str method calls necessary here?

litvinovg replied:
The added str method calls are indeed necessary.

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ProjectHasParticipantGenerator.java:
```diff
@@ -141,6 +155,37 @@ public EditConfigurationVTwo getEditConfiguration(VitroRequest vreq, HttpSession
         "?vcardName a <http://www.w3.org/2006/vcard/ns#Name> . \n" +
         "?vcardName vcard:familyName ?lastName .";
 
+    public static String getDisableVCardPrefix() {
```
This is redundant multiple times and can lead to maintainability issues among other reasons.

litvinovg replied:
I would disagree with that. Currently these methods use the same variable name, but in case someone would copy the class and try to adjust variable names it would break the form execution. So I would leave it.

wwelling replied:
Okay, but the methods return the exact same thing no matter what class members there are. Maybe the methods should be replaced with a constant?

litvinovg replied:
Replaced methods with constants.

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/edit/n3editing/configuration/generators/ProjectHasParticipantGenerator.java:
```diff
@@ -141,6 +155,37 @@ public EditConfigurationVTwo getEditConfiguration(VitroRequest vreq, HttpSession
+        return "@prefix fail_pattern: ?createVCard .\n";
+    }
+
+    public static String getDisableRealPersonPrefix() {
```
Please remove redundancy. Also, there really should be a SPARQL builder to avoid all the excessive string literals and string concatenation.

litvinovg replied:
I would disagree with that for the same reason. Currently these methods use the same variable name, but in case someone would copy the class and try to adjust variable names it would break the form execution. So I would leave it.

wwelling replied:
Okay, but the methods return the exact same thing no matter what class members there are. Maybe the methods should be replaced with a constant?

litvinovg replied:
Replaced methods with constants.

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/visualization/coauthorship/CoAuthorshipQueryRunner.java:
```diff
@@ -97,6 +97,8 @@ protected void processQuerySolution(QuerySolution qs) {
 			RDFNode documentNode = qs.get(QueryFieldLabels.DOCUMENT_URL);
 			RDFNode coAuthorURLNode = qs.get(QueryFieldLabels.CO_AUTHOR_URL);
 			RDFNode coAuthorLabelNode = qs.get(QueryFieldLabels.CO_AUTHOR_LABEL);
+			RDFNode coAuthorFamilyNode  = qs.get("coAuthorPersonFamily");
```
Please use constant.

litvinovg replied:
Fixed

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/visualization/coauthorship/CoAuthorshipQueryRunner.java:
```diff
@@ -97,6 +97,8 @@ protected void processQuerySolution(QuerySolution qs) {
 			RDFNode coAuthorURLNode = qs.get(QueryFieldLabels.CO_AUTHOR_URL);
 			RDFNode coAuthorLabelNode = qs.get(QueryFieldLabels.CO_AUTHOR_LABEL);
+			RDFNode coAuthorFamilyNode  = qs.get("coAuthorPersonFamily");
+			RDFNode coAuthorGivenNode   = qs.get("coAuthorPersonGiven");
```
Please use constant.

litvinovg replied:
Fixed

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/searchindex/extensions/VCardsAcrossContextNodes.java:
```diff
@@ -0,0 +1,455 @@
+
+                    if (stmt.getObject().isLiteral()) {
+                        switch (stmt.getPredicate().getURI()) {
+                            case "http://www.w3.org/2006/vcard/ns#familyName":
```
These should already be constants somewhere.

litvinovg replied:
Reimplemented this part

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/searchindex/extensions/VCardsAcrossContextNodes.java:
```diff
@@ -0,0 +1,455 @@
+                    Statement stmt = iter.nextStatement();
+
+                    if (stmt.getObject().isLiteral()) {
+                        switch (stmt.getPredicate().getURI()) {
```
This nesting can be reduced with a continue statement instead.

litvinovg replied:
Reimplemented this part

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/searchindex/extensions/VCardsAcrossContextNodes.java:
```diff
@@ -0,0 +1,455 @@
+            log.error(e, e);
+        }
+
+        if (!StringUtils.isEmpty(familyName)) {
```
Why not use [StringUtils.isNotEmpty](https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/StringUtils.html#isNotEmpty(java.lang.CharSequence))?

litvinovg replied:
Reimplemented this part

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/searchindex/extensions/VCardsAcrossContextNodes.java:
```diff
@@ -0,0 +1,455 @@
+    // ----------------------------------------------------------------------
+
+    private static final String LOCATE_PARTNERS_WITHOUT_RESTRICTION = ""
+            + "SELECT ?partner  \n" //
```
Isn't there a way not to have all the SPARQL in code?

litvinovg replied:
There is of course, but I am not sure of benefits for that case as our SPARQL usually coupled with code.

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/searchindex/extensions/VCardsAcrossContextNodes.java:
```diff
@@ -0,0 +1,455 @@
+            return givenName;
+        }
+
+        return null;
```
Isn't it possible to avoid the conditioning above by constructing the return value in the iteration?

litvinovg replied:
Reimplemented this part

wwelling commented on api/src/main/java/edu/cornell/mannlib/vitro/webapp/searchindex/extensions/VCardsAcrossContextNodes.java:
```diff
@@ -0,0 +1,455 @@
+            Model model = ModelFactory.createDefaultModel();
+            rdfService.sparqlConstructQuery(sparqlConstruct, model);
+
+            StmtIterator iter = model.listStatements();
```
Can the `StmtIterator` be closed with try-with-resources?

litvinovg replied:
No. It doesn't implement AutoCloseable.

wwelling commented on home/src/main/resources/rdf/i18n/fr_CA/interface-i18n/firsttime/vivo_UiLabel.ttl:
```diff
@@ -6266,3 +6266,17 @@ uil-data:degree_candidacy.VIVO
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Ne sélectionnez VCARD que si la personne n'a pas et n'est pas censée avoir un profil dans VIVO. Dans le cas contraire, choisissez le type \"Personne\"."@fr-CA ;
+        uil:hasApp       "VIVO" ;
+        uil:hasKey       "select_vcard_or_person" .
```
Missing POSIX ending newline.

litvinovg replied:
fixed

wwelling commented on home/src/main/resources/rdf/i18n/sr_Latn_RS/interface-i18n/firsttime/vivo_UiLabel.ttl:
```diff
@@ -6265,3 +6265,17 @@ uil-data:degree_candidacy.VIVO
+        rdf:type         uil:UILabel ;
+        rdfs:label       "Izaberite samo VCARD, kada osoba nema i ne bi trebalo da ima profil u VIVO-u. U suprotnom izaberite tip \"Osoba\"."@sr-Latn-RS ;
+        uil:hasApp       "VIVO" ;
+        uil:hasKey       "select_vcard_or_person" .
```
Missing POSIX ending newline.

litvinovg replied:
fixed

wwelling commented on webapp/src/main/webapp/config/listViewConfig-informationResourceInEditorship.xml:
```diff
@@ -4,16 +4,17 @@
 <list-view-config>
-    <query-select>
-        PREFIX rdfs:  &lt;http://www.w3.org/2000/01/rdf-schema#&gt;
+    <query-select>    
```
The extraneous white-space characters can be cleaned up.

litvinovg replied:
fixed

chenejac reviewed (approved):
@litvinovg it works as expected. Code looks good, as well as Serbian labels.
